Use forward slashes in ZIP files; rewrite archive tests; refine error msg

pzhlkj6612 opened the pull request:
Hello!

I'm running miniserve on Windows. I found that my files in produced ZIP files cannot be correctly extracted on macOS, but Windows Explorer didn't complain about it. After reading a post [^superuser-1382839], I thought I can fix it. Now here's the PR.

[^superuser-1382839]: windows - Zip files expand with backslashes on Linux, no subdirectories - Super User - https://superuser.com/q/1382839

I'm new to Rust. Please review my code and help me correct anything wrong.

My test. Run it on Git Bash (untested on WSL):

<details>
<summary>Bash shell script</summary>

```bash
#!/bin/bash

set -eu

# positional args
PORT=${1:-8080}

command -v miniserve 2>/dev/null || (echo 'miniserve not found' && exit 1)

playground="$(TMPDIR="$PWD" mktemp -d --suffix="-$(TZ='UTC' date '+%Y%m%dT%H%M%SZ')")"
wwwroot_dir="$playground/wwwroot"
downloaded_dir="$playground/download"
downloaded_file="$downloaded_dir/download.zip"

mkdir -p "$wwwroot_dir" "$downloaded_dir"

for name in 'out' 'out_dir/mid' 'out_dir/mid_dir/in'; do
    mkdir -p "$wwwroot_dir/${name}_dir"
    touch "$wwwroot_dir/${name}_file"
    dd if=/dev/urandom of="$wwwroot_dir/${name}_file" count=10 status=none
done

miniserve "$wwwroot_dir" --port $PORT --interfaces 127.0.0.1 --enable-zip &
pid=$!

curl -s -L "http://127.0.0.1:$PORT/?download=zip" -o "$downloaded_file"

kill $pid

unzip -l "$downloaded_file"

```

</details>

The output before this PR:

```
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  1980-01-01 00:00   wwwroot\out_dir/
     5120  1980-01-01 00:00   wwwroot\out_file
        0  1980-01-01 00:00   wwwroot\out_dir\mid_dir/
     5120  1980-01-01 00:00   wwwroot\out_dir\mid_file
        0  1980-01-01 00:00   wwwroot\out_dir\mid_dir\in_dir/
     5120  1980-01-01 00:00   wwwroot\out_dir\mid_dir\in_file
---------                     -------
    15360                     6 files

```

The output for now:

```
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  1980-01-01 00:00   wwwroot/out_dir/
     5120  1980-01-01 00:00   wwwroot/out_file
        0  1980-01-01 00:00   wwwroot/out_dir/mid_dir/
     5120  1980-01-01 00:00   wwwroot/out_dir/mid_file
        0  1980-01-01 00:00   wwwroot/out_dir/mid_dir/in_dir/
     5120  1980-01-01 00:00   wwwroot/out_dir/mid_dir/in_file
---------                     -------
    15360                     6 files

```

Before applying this fix, third-party archive utilities, for example PeaZip and 7-Zip, can't understand those backslashes:

<img height="400" alt="Screenshot of the incorrect directory layout in 3rd-party archive utility PeaZip and 7-Zip" src="https://github.com/user-attachments/assets/9dc25fb2-6e1d-4425-8826-53d0f9a4367f" />

However, Windows Explorer always shows the correct directory hierarchy despite this fix:

<img height="172" alt="Screenshot of a multi-layer directory layout in the sidebar of Windows Explorer" src="https://github.com/user-attachments/assets/d589aa29-3074-4366-b9ec-44694aeb7ecd" />

.

pzhlkj6612 committed:
Fix Windows path in ZIP files by using forward slashes only

pzhlkj6612 commented:
> Could you read into the `zip` crate and see how they deal with this?

In progress.

svenstaro commented:
Hey, I'd love to get this merged. Sorry for not giving this any attention these last few weeks. Could you please add a test for Windows so we know it won't be broken again in the future?

pzhlkj6612 commented:
No problem. Let me add necessary tests.

pzhlkj6612 committed:
Add path of failed files to error message; update comments

pzhlkj6612 committed:
Test archives with broken symlinks; test Windows path in ZIP files

pzhlkj6612 committed:
Revert changes to let test fail

pzhlkj6612 commented:
Ah-oh, workflows have not been approved. Let's wait for the Windows CI to fail, then I'll commit to revert the latest commit and you can merge it.

pzhlkj6612 committed:
Fix failed TAR test - directory walking result varies by OS

svenstaro commented on tests/archive.rs:
```diff
@@ -90,3 +92,67 @@ fn archives_are_disabled_when_indexing_disabled(
+fn zip_archives_store_entry_name_in_unix_style(
+    #[with(&["--enable-zip"])] server: TestServer,
+) -> Result<(), Error> {
+    let body = reqwest::blocking::get(server.url().join("someDir/?download=zip")?)?;
```
Please use reqwest like [so](https://github.com/svenstaro/miniserve/blob/45b4e26d2ad9f5ea0318adf7b52b245779d6113f/tests/api.rs#L27).

svenstaro commented on tests/archive.rs:
```diff
@@ -90,3 +92,67 @@ fn archives_are_disabled_when_indexing_disabled(
+    #[with(&["--enable-tar-gz", "--enable-tar", "--enable-zip"])] server: TestServer,
+) -> Result<(), Error> {
+    let download_archive = |download_type: &str| {
+        let body = reqwest::blocking::get(
```
Please use reqwest like [so](https://github.com/svenstaro/miniserve/blob/45b4e26d2ad9f5ea0318adf7b52b245779d6113f/tests/api.rs#L27).

svenstaro commented:
It looks like this currently only contains the tests but not the fixes.

pzhlkj6612 committed:
Merge "master"; use reqwest_client instead

pzhlkj6612 committed:
Shrink imports; fix English typo; refine "test_tar_archives"

pzhlkj6612 committed:
Fix Windows path in ZIP files by using forward slashes only

Re-apply ad8d3d9a4862cb1e8b139132920273470e1cb991 by reverting a4289ffd637a836e6224950bd59e2995e2c70fcc

pzhlkj6612 committed:
Apply review suggestions

- sort imports
- clean dependency
- refine comments and English
- use long parameter

Co-authored-by: Sven-Hendrik Haase <svenstaro@gmail.com>

pzhlkj6612 committed:
Rewrite archive tests with throughout coverage

- Universal definitions.
- Reuse HTTP client.
- Docstring.

svenstaro commented on src/archive.rs:
```diff
@@ -234,39 +234,63 @@ where
+
+            // Workaround for Windows path in ZIP files:
+            // Always use forward slashes for all paths to avoid literal backslashes in saved entry path.
+            // TODO: remove this when the "zip" cargo has the ability to process a directory as a whole.
```
According to https://github.com/zip-rs/zip2/issues/376 and the [release notes of zip v7.2.0](https://github.com/zip-rs/zip2/releases/tag/v7.2.0), this is now handled by upstream. At least I see they now have tests handling this. Could you check out whether that now does the job? It's a bit odd though that they have that issue still open.

svenstaro commented on tests/archive.rs:
```diff
@@ -1,129 +1,269 @@
+        assert!(
+            contains_text(document, text),
+            "Expected link text '{}' to be present",
+            text
```
```suggestion
            "Expected link text '{text}' to be present",
```

svenstaro commented on tests/archive.rs:
```diff
@@ -1,129 +1,269 @@
+        assert!(
+            !contains_text(document, text),
+            "Expected link text '{}' to be absent",
+            text
```
```suggestion
            "Expected link text '{text}' to be absent",
```

svenstaro reviewed (approved):
Tiny nit but apart from that I really like these tests. I was at first a bit skeptical about the `ArchiveKind` thing as it adds some complexity but I think here it actually makes sense in the context of the `archives_links_and_downloads` test.

svenstaro commented:
Happy to merge after you figure out whether upstream's fix works for us. :)

pzhlkj6612 committed:
Merge "master"

pzhlkj6612 committed:
Apply code review suggestions

The 2nd parameter of "assert!()" has "format!()" integrated.

Co-authored-by: Sven-Hendrik Haase <svenstaro@gmail.com>

pzhlkj6612 committed:
Revert "using forward slashes in ZIP on Windows" and a related test

Our dependency "zip" has fixed a bug about setting local platform in the ZIP since v7.2.0: https://github.com/zip-rs/zip2/commit/9e9badde0510a863b2362affc73eed32da3fbe5c, so we don't need to fix it here.

This effectively reverts the main part of commits:

- The fix: ad8d3d9a4862cb1e8b139132920273470e1cb991 (the origin) or 3b59a69bdca2b0672bf5c55a935cfa928aec4dce (the re-applied one)
- The test: 90c523425e092c7137c5450c8276167954e64ff4

pzhlkj6612 committed:
Reapply "using forward slashes in ZIP on Windows" and a related test

macOS: rememeber, no forward slashes.

This reverts commit a06e31e716bede3800d3f328a0322d9b164a1079.

svenstaro commented on src/archive.rs:
```diff
@@ -234,39 +234,63 @@ where
+                    .trim_end_matches(r"\") // every branch ends with two backslashes "\\".
+                    .replace(r"\", "/"); // every branch uses backslash "\" as path separators.
+                let leaf = current_entry_name.to_string_lossy();
+                format!("{branch}/{leaf}") // construct a Unix-style path in the simplest way.
```
Shouldn't the comment say "always use forward slashes"?

pzhlkj6612 replied:
1fe616742935664d0e23c8b9886277b1bec4a280

svenstaro commented on tests/archive.rs:
```diff
@@ -1,129 +1,267 @@
+    assert_eq!(resp.status(), expected);
+
+    Ok(Document::from_read(resp)?)
+}
```
I think it would probably make sense to make this a universal utility and then use it everywhere? We have sooo many `Document::from_read`s in all the tests that this might be a great follow-up to tidy up some test code if you're up for it. At any rate, this should probably go into `tests/utils/mod.rs` for now.

pzhlkj6612 committed:
Fix wrong comments about slashes

oops

Co-authored-by: Sven-Hendrik Haase <svenstaro@gmail.com>
